lib/libc/gen/sysctl.3
.\" Copyright (c) 1993 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" %sccs.include.redist.roff%
.\"
.\"	@(#)sysctl.3	6.1 (Berkeley) 3/2/93
.\"
.Dd ""
.Dt SYSCTL 2
.Os
.Sh NAME
.Nm sysctl
.Nd get or set kernel state
.Sh SYNOPSIS
.Fd #include <sys/sysctl.h>
.Ft int
.Fn sysctl "int *name" "u_int namelen" "void *old" "u_int *oldlenp" "void *new" "u_int newlen"
.Sh DESCRIPTION
The
.Fn sysctl
function retrieves kernel state and allows processes with
appropriate privilege to set kernel state.
The state to be set is described using a ``MIB'' style name,
described in a
.Fa namelen
length array of integers pointed to by
.Fa name .
The top level names are defined with a CTL_ prefix in
.Pa <sys/sysctl.h> .
The next levels down are found in the files given in the ``FILES''
section of this manual page.
.Pp
To retrieve a value,
.Fa old
is set to point to a buffer
into which the requested value is to be placed.
The length of the buffer is given in the location pointed to by
.Fa oldlenp .
The size of the returned value is put in the location pointed to by
.Fa oldlenp .
The size of the requested value can be determined by calling 
.Fn sysctl
with a NULL parameter for
.Fa old ;
the size of the value will be returned in the location pointed to by
.Fa oldlenp .
If the old value is not desired,
.Fa old
and
.Fa oldlenp
can be set to NULL.
.Pp
To set a new value,
.Fa new
is set to point to a buffer of length
.Fa newlen
from which the requested value is to be taken.
If the setting of a new value is not desired,
.Fa new
should be set to NULL and
.Fa newlen
set to 0.
.Pp
For example, to retrieve the maximum number of processes allowed
in the system, one would use the follow request:
.sp
.Bd -literal -offset indent -compact
int name[2], len, maxproc;

name[0] = CTL_KERN;
name[1] = KERN_MAXPROC;
len = sizeof maxproc;
sysctl(name, 2, &maxproc, &len, NULL, 0);
.Ed
.Sh RETURN VALUES
If the call to
.Fn sysctl
is successful, the length of the old value is returned.
Otherwise \-1 is returned and
.Va errno
is set appropriately.
.Sh ERRORS
The following error may be reported:
.Bl -tag -width Er
.It Bq Er EFAULT
The buffer
.Fa name ,
.Fa old ,
.Fa new ,
or length pointer
.Fa oldlenp
contains an invalid address.
.It Bq Er EINVAL
The
.Fa name
array is less than two or greater than CTL_MAXNAME.
.It Bq Er EINVAL
A non-null
.Fa new
is given and its specified length in
.Fa newlen
is too large or too small.
.It Bq Er ENOMEM
The length pointed to by
.Fa oldlenp
is too short to hold the requested value.
.It Bq Er ENOTDIR
The
.Fa name
array specifies an intermediate rather than terminal name.
.It Bq Er EOPNOTSUPP
The
.Fa name
array specifies a value that is unknown.
.It Bq Er EPERM
An attempt is made to set a read-only value.
.It Bq Er EPERM
A process without appropriate privilege attempts to set a value.
.El
.Sh FILES
.Bl -tag -width <vm/vm_param.h> -compact
.It Pa <sys/sysctl.h>
definitions for top level identifiers and second level kernel
and hardware identifiers
.It Pa <sys/socket.h>
definitions for second level network identifiers
.It Pa <vm/vm_param.h>
definitions for second level virtual memory identifiers
.El
.Sh SEE ALSO
.Xr sysctl 8
.Sh HISTORY
.Fn sysctl
first appeared in 4.4BSD.


usr.sbin/sysctl/sysctl.c
/*
 * Copyright (c) 1993 The Regents of the University of California.
 * All rights reserved.
 *
 * %sccs.include.redist.c%
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1993 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)sysctl.c	5.1 (Berkeley) 3/30/93";
#endif /* not lint */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/sysctl.h>
#include <sys/socket.h>
#include <vm/vm_param.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

char *topname[] = CTL_NAMES;
char *kernname[] = CTL_KERN_NAMES;
char *vmname[] = CTL_VM_NAMES;
char *netname[] = CTL_NET_NAMES;
char *hwname[] = CTL_HW_NAMES;

struct list {
	char	**list;
	int	size;
} secondlevel[] = {
	{ 0, 0 },			/* CTL_UNSPEC */
	{ kernname, KERN_MAXID },	/* CTL_KERN */
	{ vmname, VM_MAXID },		/* CTL_VM */
	{ 0, 0 },			/* CTL_FS */
	{ netname, NET_MAXID },		/* CTL_NET */
	{ 0, 0 },			/* CTL_DEBUG */
	{ hwname, HW_MAXID },		/* CTL_HW */
	{ 0, 0 },			/* CTL_MACHDEP */
};

int	Aflag, aflag, wflag;

int
main(argc, argv)
	int argc;
	char *argv[];
{
	extern char *optarg;
	extern int optind;
	int ch;

	while ((ch = getopt(argc, argv, "Aaw")) != EOF) {
		switch (ch) {

		case 'A':
			Aflag = 1;
			break;

		case 'a':
			aflag = 1;
			break;

		case 'w':
			wflag = 1;
			break;

		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (Aflag || aflag) {
		listall();
		exit(0);
	}
	if (argc == 0)
		usage();
	while (argc-- > 0)
		parse(*argv, 1);
	exit(0);
}

/*
 * List all variables known to the system.
 */
listall()
{
	struct list *lp;
	int lvl1, lvl2;
	char *cp, name[BUFSIZ];

	for (lvl1 = 1; lvl1 < CTL_MAXID; lvl1++) {
		lp = &secondlevel[lvl1];
		if (lp->list == 0)
			continue;
		strcpy(name, topname[lvl1]);
		cp = &name[strlen(name)];
		*cp++ = '.';
		for (lvl2 = 1; lvl2 < lp->size; lvl2++) {
			strcpy(cp, lp->list[lvl2]);
			parse(name, Aflag);
		}
	}
}

/*
 * Parse a name into a MIB entry.
 * Lookup and print out the MIB entry if it exists.
 * Set a new value if requested.
 */
parse(string, flags)
	char *string;
	int flags;
{
	int indx, size;
	int isclockrate = 0;
	void *newval = 0;
	int intval, newsize = 0;
	struct list top, *lp;
	int mib[CTL_MAXNAME];
	char *cp, *bufp, buf[BUFSIZ], strval[BUFSIZ];

	bufp = buf;
	snprintf(buf, BUFSIZ, "%s", string);
	if ((cp = strchr(string, '=')) != NULL) {
		if (!wflag) {
			fprintf(stderr, "Must specify -w to set variables\n");
			exit(2);
		}
		*strchr(buf, '=') = '\0';
		*cp++ = '\0';
		while (isspace(*cp))
			cp++;
		if (isdigit(*cp)) {
			intval = atoi(cp);
			newval = &intval;
			newsize = sizeof intval;
		} else {
			newval = cp;
			newsize = strlen(cp);
		}
	}
	top.list = topname;
	top.size = CTL_MAXID;
	if ((indx = findname(string, "top", &bufp, &top)) == -1)
		return;
	mib[0] = indx;
	lp = &secondlevel[indx];
	if (lp->list == 0) {
		fprintf(stderr, "%s: class is not implemented\n",
		    topname[indx]);
		return;
	}
	if ((indx = findname(string, "second", &bufp, lp)) == -1)
		return;
	mib[1] = indx;
	if (bufp) {
		fprintf(stderr, "name %s in %s is unknown\n", *bufp, string);
		return;
	}
	switch (mib[0]) {

	case CTL_KERN:
		switch (mib[1]) {
		case KERN_VNODE:
		case KERN_FILE:
			if (flags == 0)
				return;
			fprintf(stderr,
			    "Use pstat to view %s information\n", string);
			return;
		case KERN_PROC:
			if (flags == 0)
				return;
			fprintf(stderr,
			    "Use ps to view %s information\n", string);
			return;
		case KERN_CLOCKRATE:
			isclockrate = 1;
			break;
		}
		break;

	case CTL_HW:
		break;

	case CTL_VM:
		if (mib[1] == VM_LOADAVG) {
			double loads[3];

			getloadavg(loads, 3);
			fprintf(stdout, "%s: %.2f %.2f %.2f\n", string,
			    loads[0], loads[1], loads[2]);
			return;
		}
		if (flags == 0)
			return;
		fprintf(stderr,
		    "Use vmstat or systat to view %s information\n", string);
		return;

	case CTL_NET:
		if (flags == 0)
			return;
		fprintf(stderr, "Use netstat to view %s information\n", string);
		return;

	case CTL_FS:
	case CTL_DEBUG:
	case CTL_MACHDEP:
		break;

	default:
		fprintf(stderr, "Illegal top level value: %d\n", mib[0]);
		return;
	
	}

	size = BUFSIZ;
	if (sysctl(mib, 2, buf, &size, newsize ? newval : 0, newsize) == -1) {
		if (flags == 0)
			return;
		switch (errno) {
		case EOPNOTSUPP:
			fprintf(stderr, "%s: value is not available\n", string);
			return;
		case ENOTDIR:
			fprintf(stderr, "%s: specification is incomplete\n",
			    string);
			return;
		case ENOMEM:
			fprintf(stderr, "%s: type is unknown to this program\n",
			    string);
			return;
		default:
			perror(string);
			return;
		}
	}
	if (isclockrate) {
		struct clockinfo *clkp = (struct clockinfo *)buf;

		fprintf(stdout,
		    "%s: hz = %d, tick = %d, profhz = %d, stathz = %d\n",
		    string, clkp->hz, clkp->tick, clkp->profhz, clkp->stathz);
		return;
	}
	if (size == sizeof(int) && !isprint(buf[0]))
		if (newsize == 0)
			fprintf(stdout, "%s = %d\n", string, *(int *)buf);
		else
			fprintf(stdout, "%s: %d -> %d\n", string, *(int *)buf,
			    *(int *)newval);
	else
		if (newsize == 0)
			fprintf(stdout, "%s = %s\n", string, buf);
		else
			fprintf(stdout, "%s: %s -> %s\n", string, buf, newval);
	return;
}

/*
 * Scan a list of names searching for a particular name.
 */
findname(string, level, bufp, namelist)
	char *string;
	char *level;
	char **bufp;
	struct list *namelist;
{
	char *name;
	int i;

	if (namelist->list == 0 || (name = strsep(bufp, ".")) == NULL) {
		fprintf(stderr, "%s: incomplete specification\n", string);
		return (-1);
	}
	for (i = 0; i < namelist->size; i++)
		if (!strcmp(name, namelist->list[i]))
			break;
	if (i == namelist->size) {
		fprintf(stderr, "%s level name %s in %s is invalid\n",
		    level, name, string);
		return (-1);
	}
	return (i);
}

usage()
{

	(void)fprintf(stderr, "usage:\t%s\n\t%s\n\t%s\n",
	    "sysctl [-w] variable ...", "sysctl -a", "sysctl -A");
	exit(1);
}
